### PR TITLE
Three component

### PR DESCRIPTION
**Pull request by grischaerbe**
Grab a coffee, this might take a while.

### The _Why_

As it's a common criticism of three.js component libraries that they're just _wrapping_ things and can't keep up with three.js changes, I decided to add a component that mimics what `react-three-fiber` does with a custom react renderer.

It's a component that uses the rigid naming conventions of three.js and allows you to use **anything** that three.js offers as a svelte component.

### Example

Let's have a look at an example:

```html
<Three type="Mesh" scale.z={4} let:object>
	<InteractiveObject {object} interactive on:click={onClick} />
	<Three type="MeshStandardMaterial" flatShading opacity={0.7} transparent side={DoubleSide}>
		<Three type="Color" attach="color" args={['hotpink']} />
	</Three>
	<Three type="TorusGeometry" args={[1.2, 0.3, 70, 36]} />
</Three>
```

This example adds a `THREE.Mesh` to the scene graph and uses [slot props](https://svelte.dev/tutorial/slot-props) to be able to pass the actual instance to nested components such as the _trait component_  `<InteractiveObject />`. Then, the same component `<Three />` is used to add a material and a geometry to the mesh. Nested in that material is another `<Three />`, that assigns a `THREE.Color` to be the color of the property `color` of the material. Also note that properties on the objects can be used to set object properties reactively.

### How it works

Currently, what `<Three />` does is:

- check what has been passed to the property `type` and tries to instantiate it (if it's a constructor) – \*_Currently only constructors are allowed, but this restriction could be weakened I guess_.
- Makes assumptions of the resulting instance:
  - Is it an object that extends `THREE.Object3D`? Then it probably needs to be added to its parent (via the trait component `<SceneGraphObject>`) and removed `onDestroy`.
  - It's not? Then it gets _attached_. Attaching means that it will be bound to a property of its parent. In the case of `<Three type="Color" args={['orange']} attach="color" />` this means that the resulting `THREE.Color` will be attached to the property `color` of its parent, which is a `THREE.MeshStandardMaterial`. Materials and geometries don't need to be explicitly attached as we can safely assume that all material names end with `Material` and all geometry names end with `Geometry` and therefore they attach to their respective property on the parent: `material` or `geometry`.
- All props except `type`, `object`, `args` and `attach` can be used to set properties on the object (see `flatShading` in the example above).
- (Currently instantiating a camera (e.g. `<Three type="PerspectiveCamera" />`) always sets the camera as the default one, this needs more control.

### Attaching

The property `attach` can be used to attach an object to a parent property. `attach` can be a string …

```html
<Three type="MeshStandardMaterial">
	<Three type="Color" args={['orange']} attach="emissive" />
</Three>
```

… with dot-notation to access nested properties … 

```html
<Three type="Mesh">
	<Three type="BufferGeometry">
		<Three type="BufferAttribute" attach="attributes.position" />
	</Three>
</Three>
```

… or a function that runs on initialization and `onDestroy`:

```html
<Three
	type="MeshStandardMaterial"
	roughness={0.3}
	metalness={1}
	envMap={renderTarget.texture}
	attach={(parent, self) => {
		parent.material = self
		return () => {
			parent.material = null
		}
	}}
/>
```

### Props

- All properties on the resulting instance are exposed as props and typed as best as possible:

```html
<Three type="MeshStandardMaterial" flatShading opacity={0.7} transparent side={DoubleSide} />
```

- If a `set()` function exists on the property (such as the `color.set()` function on a `THREE.Material`) you can pass whatever you can pass in that set function as the sole value of the corresponding prop:

```html
<Three type="MeshStandardMaterial" color="green" />
```

- You can use dot-notation on props to pierce into nested properties:

<Three type="Mesh" position.x={4} />

- All props except `args` **are reactive**:

```html
<script>
	…
	let color = 'red'
</script>

<Three type="Mesh" scale.z={4} let:object>
	<Interaction
		{object}
		interactive
		on:click={() => {
			color = color === 'red' ? 'blue' : 'red'
		}}
	/>
	<Three
		type="MeshStandardMaterial"
		{color}
		flatShading
		opacity={0.7}
		transparent
		side={DoubleSide}
	/>
	<Three type="TorusGeometry" args={[1.2, 0.3, 70, 36]} />
</Three>
```

### Type Support

As good as I can make it, but as most stuff is inferred, it's more than wonky from time to time.

Bildschirmfoto 2022-10-29 um 00 45 20

### CODE REVIEW

**Comment by grischaerbe** [comment]
With the current prop setting memoization, it will lead to an error if the prop type of e.g. `scale` changes from `number` to `[number]`. Prop setter memos should not include prop type checking but only path resolving.

**Comment by DefinitelyMaybe** [comment]
Interesting. It seems like this would allow us to wrap anybody's three.js extension/addon (i.e. gkjohnson's  [mesh-bvh](https://github.com/gkjohnson/three-mesh-bvh), [three-gpu-pathtracer](https://github.com/gkjohnson/three-gpu-pathtracer), [three-bvh-csg](https://github.com/gkjohnson/three-bvh-csg) ?)

Also we _could_ reduce a lot of threlte to this one component but surely we still have both? Would they work together just fine?

How would the documentation go? If there's already a svelte component, use that first?

Would threlte/rapier work with this?

> It's a component that uses the rigid naming conventions of three.js and allows you to use anything that three.js offers as a svelte component.

So we _never_ have to play catch up. Everyone would be able to develop at the speed of three.js. We'd be a multiplier on-top not an overhead.

**Review by PontusHorn — COMMENTED** [review]
Looked through the code - nice work! This solution seems neat and understandable to me, despite not being super familiar with the library's internals. Left a few comments mostly around naming and stuff like that, but I realise all that stuff might be up in the air anyway, so take them or leave them.

**Comment by grischaerbe** [comment]
Thanks @PontusHorn for your review! I guess this all still needs a fair amount of refactoring ;)

File: packages/core/src/lib/three/Three.svelte [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  type $$Props = {
+    js: T
+    args?: Args | any[]
+    attach?: string | ((parent: any, self: any) => (() => void) | void)
+    object?: Object
+  } & Record<
+    string,
+    any
+  > & {
```
[the remark this answers is not in the document]
Comment (grischaerbe) [reply packages/core/src/lib/three/Three.svelte:46]: True! That is needed for pierced props, such as `position.x={3}`. Good point!

**Comment by grischaerbe** [comment]
This last commit is mixing things up a little.
Instead of a string provided to `type` that determines the import from `three`, a value (class, primitive, anything) can be provided to the component. If it's a class, the prop `args` will be used to instantiate it, all other values will be left as-is.
By that, tree-shaking works out-of-the-box, and the type ergonomics are way better for imports outside the `THREE` namespace. Also, there's no need to `extend` the imports anymore since everything has to be imported manually anyway.
The component now does a very bad job at performance optimizations when it comes to applying props but I think that's not really important right now.

**Comment by grischaerbe** [comment]
# RFC

So, I talked to another developer whose opinion I highly value. We ended up with different questions to which I don't have clear answers:

## To `args` or not to `args`

`react-three-fiber` transforms `<mesh />` to a `THREE.Mesh` Instance by just looking at the tag name. Without a preprocessor we cannot do that. As such we have to rely on a regular component which is the original topic of this PR. `react-three-fiber` uses the prop `args` to pass arguments to a constructor: `<boxGeometry args={[1, 1, 1]} />` → `new THREE.BoxGeometry(1, 1, 1)`.
Because we don't have the comfort of `<boxGeometry>` and have to stick to a regular component (such as `<Three />`) we need to pass a class definition *or* an instance of a class (or basically any other value) to that component to let it know what to do with it. 

### With `args`

Now to mimic `react-three-fiber`'s approach, we would:
- `<Three type={BoxGeometry} />` to instantiate a `THREE.BoxGeometry` with no constructor parameters
- `<Three type={BoxGeometry} args={[1, 1, 1]} />` or instantiate a `THREE.BoxGeometry` with constructor parameters

Pro:
- Looks **Magic**
- Less to type if no args need to be passed
- Close to `react-three-fiber`

Con:
- More to type *if* args have to be passed
- Less transparent for beginners
- Types need to be inferred from class definitions *and* class instances. `args` cannot be typed properly due to [the utility type `ConstructorParameters<T>` not being able to pick up overloads](https://github.com/microsoft/TypeScript/issues/37079) and therefore needs a fallback (`any[]`) which does not meet my regular type standards.
- Let's assume that the prop to pass a class definition or a class instance is the same, then we get naming conflicts:
  - `<Three type={BoxGeometry} />` for a class definition
  - `<Three type={new BoxGeometry(1, 1, 1)} />` for a class instance
- (Personally I don't like to pass arguments as an array: it looks odd and types more oddly)

### Without `args`

Because we have to pass `BoxGeometry` as a prop anyway, there's little to no reason to just pass an instance in the first place:
- `<Three type={new BoxGeometry()} />`
- `<Three type={new BoxGeometry(1, 1, 1)} />`

Pro:
- Intuitive
- Less to type if you pass arguments
- Easier to infer types
- Constructor overloads are not a problem → type safety is not a threlte concern
- Prop name does not need to reflect class *definitions* and class *instances* (but also every other value, so take that with a grain of salt)

Con:
- More semantic noise (`new`, `new`, `new`, …)
- Less close to `react-three-fiber`
- Looks less *magic*

## The prop name `type`

This is heavily depending on the question before, so let's split this up again:

### With `args`

Currently, `<Three>` accepts the property `type` and – if a class definition is provided – the constructor arguments as the prop `args`. Let's focus on the naming of whatever `type` is.

Here's a list of what could end up in that prop:
- class definitions (`THREE.BoxGeometry`)
- class instances (`new THREE.BoxGeometry()`)
- regular objects or arrays (`{ x: 5, z: 8 }`)
- scalar values (`true`)

Of course, the hunt for a prop name gets harder when class definitions are part of the pack. Currently we settled for `type`, here's a list of other candidates:

- `type` – `<Three type={BoxGeometry} />`
- `js` – `<Three js={BoxGeometry} />`
- `this` – `<Three this={BoxGeometry} />`
- `as` – `<Three as={BoxGeometry} />`
- `use` – `<Three use={BoxGeometry} />`

### Without `args`

Marginally easier because class *definitions* do not need to be reflected in the prop name anymore. The prop name `use` might be generic but expressive enough:

- `<Three use={new BoxGeometry()} />`

## Semantic Noise

From now on let's assume we stick with **no `args` and `use` as the magic prop name**.

Let's have a look at a component tree from the upcoming demo (as a screenshot because GitHub does not support Svelte highlighting yet):

Bildschirmfoto 2022-11-08 um 00 20 31

It's hard to quickly skim over a document and spot particular elements. This is a react-style counterpart:

Bildschirmfoto 2022-11-08 um 00 23 32

The difference in readability and overview is huge.

### How to deal with semantic noise

- **Shorten the component name substantially:**
  - <details><summary>Screenshot</summary>Bildschirmfoto 2022-11-08 um 00 29 01</details>
  - (What I kind of like about this one is that `T` is often used as a type placeholder in generics, kind of what this component is about, but looking at it now it's not what I'd like to look at day in, day out.)

- **Semantically divide concerns to separate components although they do technically the same:**
  - <details><summary>Screenshot</summary>Bildschirmfoto 2022-11-08 um 00 32 17</details>
  - (This would also mean using different prop names for each concern, but that's for a later date.)

- **Just deal with it and embrace the benefits of this approach:**
  - Types
  - Tree-shaking
  - No need to `extend` a catalogue

---

I'd love to hear your comments on these topics ✌️

**Comment by DefinitelyMaybe** [comment]
Copy r3f and go with args. Less mental overhead for both groups, easier conversion when needed.

Partly I thought we were going down the route of: use this component when there isn't anything. If we did suddenly move to only this component then readability/semantic noise would be a problem as demonstrated but encapsulated within other svelte components would be the way to go

**Comment by grischaerbe** [comment]
I just did a small performance benchmark: In a very naive and simple test the "rendered" component is about 2-3x faster than the "wrapped" counterpart. The test was rendering 2000 cubes and translating it one axis on every frame. Stats are taken with [stats.js](https://github.com/mrdoob/stats.js/) and accounted for the whole [frameloop](https://github.com/threlte/threlte/blob/207d6835b91a513db2100bf9cf267ec994cf4283/packages/core/src/lib/lib/frameloop.ts#L66-L92) (running frame handlers, rendering). Very promising results!

|  | Wrapped | Rendered |
|---|---|---|
| fps | ~18 | ~54 |
| memory (MB) | ~145 | ~95 |
| frametime (ms) | ~6.5 | ~5.5 |

**Wrapped component**

```html
<script lang="ts" context="module">
	const geometry = new BoxGeometry()
	const material = new MeshBasicMaterial()
</script>

<script lang="ts">
	import { Mesh, useFrame } from '@threlte/core'
	import { BoxGeometry, MeshBasicMaterial } from 'three'

	let positionX = 0

	useFrame(() => {
		positionX = Math.sin(Date.now() / 1000 + Math.random() * 10) * 2
	})
</script>

<Mesh {geometry} {material} position={{ x: positionX }} />
```

**Rendered component**

```html
<script lang="ts" context="module">
	const geometry = new BoxGeometry()
	const material = new MeshBasicMaterial()
</script>

<script lang="ts">
	import { Three2, useFrame } from '@threlte/core'
	import { MeshBasicMaterial } from 'three'
	import { BoxGeometry } from 'three'
	import { Mesh } from 'three'

	let positionX = 0

	useFrame(() => {
		positionX = Math.sin(Date.now() / 1000 + Math.random() * 10) * 2
	})
</script>

<Three2 type={Mesh} position.x={positionX}>
	<Three2 type={geometry} />
	<Three2 type={material} />
</Three2>
```

**Comment by MartinMikusat** [comment]
I mentioned I'm not a big fan of this on Discord and I've managed to gather my thoughts about this into something more coherent. 

I'm working on a major project where 3D is a ultimately a small portion of a much larger whole. Right now, Threlte feels like writing Svelte, I just get 3D back. This change doesn't fit into the way we usually write Svelte - all this attaching instead of using props, importing from Three.js and adding that as a prop to some "Three" svelte component. 

To me, that introduces another context switch into my workflow. It's a small thing but it adds up. Instead of staying in the "Svelte" context, I will have to switch between "Svelte" and "Threlte" context which is just another layer of friction that I would personally like to avoid.

That said, I trust your judgement and if this is providing major improvement to the performance I would ultimately prefer that over better DX, even though I won't like it. I appreciate Threlte even exist as it's going to provide major productivity boost either way.

**Comment by Rolikasi** [comment]
I like threltes DX now. It's easy to read and understand. 
```
<Canvas>
   <PerspectiveCamera position={{ x: 20, y: 20, z: 20 }}>
	<OrbitControls
	maxPolarAngle={DEG2RAD * 80}
	enableZoom={true}
	target={{ y: 0.5 }}
	maxDistance={100}
	minDistance={20}
			/>
   </PerspectiveCamera>
   <Group scale={$scale}>
      <Mesh
	interactive
	rotation={{ x: 90 * (Math.PI / 180) }}Ï
	scale={0.1}
	position={{ y: 0.5 }}
	geometry={svgToGeometry(svg)}
        material={new MeshStandardMaterial({ color: '#333333' })}
	/>
   </Group>
</Canvas>
```
I'm gonna be a bit frustrated always using <Three /> components and navigating in that noise. 
But if it's gonna improve performance, size, and ability to add anything from three js, it might be worth it.

**Comment by grischaerbe** [comment]
Is there a Type Wizard around here who can help inject types for the preprocessor solution?

**Comment by DefinitelyMaybe** [comment]
> I just did a small performance benchmark: In a very naive and simple test the "rendered" component is about 2-3x faster than the "wrapped" counterpart. The test was rendering 2000 cubes and translating it one axis on every frame. Stats are taken with [stats.js](https://github.com/mrdoob/stats.js/) and accounted for the whole [frameloop](https://github.com/threlte/threlte/blob/207d6835b91a513db2100bf9cf267ec994cf4283/packages/core/src/lib/lib/frameloop.ts#L66-L92) (running frame handlers, rendering). Very promising results!
> 
> Wrapped	Rendered
> fps	~18	~54
> memory (MB)	~145	~95
> frametime (ms)	~6.5	~5.5

Oh my, ok second option it is, this becomes the defacto. To clarify, we would or wouldn't re-write the current components with it and exported as a convenience?

Hmm, the doc's might become all about snippets and how-to's which would be nice.

**Comment by grischaerbe** [comment]
I still don't have a clue what to do with the existing components. I'm quite happy with the preproccessor-based approach and tried to implement types via TS ambient types. Generally it works but I couldn't make it work with THREE types, let's see what tomorrow brings 🙂

**Comment by PontusHorn** [comment]
If you want another pair of eyes on the typing issues I'd be happy to take a crack at it tonight.

**Comment by grischaerbe** [comment]
> If you want another pair of eyes on the typing issues I'd be happy to take a crack at it tonight.

Sure, please go ahead! For testing purposes i added the preprocessor to apps/docs but ultimately it should go in `@threlte/core` (because its hard dependency is the `<Three>` component). The component `<Three>` uses generics and i thought we could use that to drive the ambient types:

```ts
// this does not work!
type ThreeNamespace = import('three')
declare namespace T {
  [Key in ThreeNamespace]: ThreeComponentDef<ThreeNamespace[Key]>
}
```

In order to be able to type this:

```
<T.MeshStandardMaterial />
```

Which the preprocessor would the transform to this:

```
// imports added
<Three type={MeshStandardMaterial} />
```

But I have little experience in authoring ambient types and somehow they don't work like that.

**Comment by grischaerbe** [comment]
Okay, so even though this is marked as a draft PR I feel sorry for all the force pushing and shifting around of versions of all kinds of packages. SvelteKit and its breaking changes were giving me a hard time. I decided to not touch SvelteKit and Vite deps until it's really 1.0. Until then I locked them.

The good news is that the deployment finally works and the [preprocessor is working](https://threlte-deployment-test-qlfjkpbyg-threlte.vercel.app/preprocess). Now on to the types!

**Comment by grischaerbe** [comment]
## Types for a preprocessed component

There are several hurdles to take:

First there are two routes to go

**a)** mimic a web component: `<t.Mesh />`
  - `bind:ref` will cause errors that I think are hard to intercept. The svelte extension naturally thinks that this is a web component.
  
**b)** mimic a regular component : `<T.Mesh />`
  - I *think* we should be possible to introduce ambient types. The Svelte extension is still complaining about a "missing declaration" and introduces annoying yellow underlines but TypeScript stops complaining at that point.

I think **b)** is definitely the way forward.

So let's say we go with b), we have two options again:
**i)** Ambient types via a `.d.ts` file, no import in consuming component
**ii)** Explicitly imported types: use a single `import { T } from '@threlte/core'` in consuming component

I'm considering b.ii) because it's closer to importing a regular component. Also type authoring is easier in a regular type than an ambient type from my experience. Now here's the tricky part that we can actually not make that a pure type export but we should actually export something so that you can also use it without `<script lang="ts">` but I didn't test that. Curious for your input!

**Comment by DefinitelyMaybe** [comment]
Here and still reading.

Would either option work with custom three classes?

**Comment by DefinitelyMaybe** [comment]
So basically would there ever be a circumstance where someone's creating a new component with this new `<Three>` component and **not** want the THREE types to come along too?

i.e. we don't need to always include the types if everyone's always using them but if there's a use case where someone doesn't want them then they'll always get in the way.

yes?

**Comment by grischaerbe** [comment]
## Finishing touches

### `<Three>`

### Events

The component now listens to any event specified on the consuming side and adds respective event listeners: `<Three type={OrbitControls} on:change={(e) => console.log(e)} />` will run `addEventListener('change', …)` on the instance of the `OrbitControls` and forward all events to the handler passed to `on:change`. Unfortunately the three.js types are lacking event names, so you have to stick to the three.js documentation for event names.

## Types

The types for the component `<Three>` and the types for the preprocessed component `<T>` are shared, so type changes are made in only one place.

## Preprocessor

- The preprocessors default Prefix is now `T`.
- The preprocessor now also strips the import of `T`: `import { T, useThrelte } from '@threlte/core'` → `import { useThrelte } from '@threlte/core'`.
- Currently the preprocessor relies on `svelte.parse`, so all preprocessors **must** run beforehand, to e.g. transform TypeScript to JavaScript. This is done via [`svelte-sequential-preprocessor`](https://github.com/pchynoweth/svelte-sequential-preprocessor). I think it can be refactored to not rely on `svelte.parse`.

All in all I'm quite happy with the result so far. Take a look at this component structure for example:

Bildschirmfoto 2022-11-11 um 18 08 53

In my opinion this now provides a very good DX as well as flexibility and type safety. 

Next steps:
- Refactor preprocessor
- Documentation (big one)
- *SHIP IT* 🚀

**Pull request merged**